File: Week 5/solutions/D/D_password.cpp
#include <bits/stdc++.h>
using namespace std;


const int INF = 1e9 + 5;

int main(){
    // There are a lot of approaches to this problem; the constraints are small enough
    // that you can iterate over all the palindromes to find the smallest. But it becomes
    // simpler to write if we just precompute them, and use a set to find the smallest
    set<int> all_palindromes;
    for(int i = 100; i <= 999; i++){ // No leading zeroes; start at 100
        string s = to_string(i);
        string t = s;
        reverse(t.begin(), t.end());
        s += t;
        all_palindromes.insert(stoi(s));
    }
    int qr;
    cin >> qr;
    while(qr--){
        int k;
        cin >> k;
        auto it = all_palindromes.upper_bound(k);
        int best = INF, best_palindrome = -1;
        if(it != all_palindromes.end()){ // Check palindrome greater than k
            best = *it - k;
            best_palindrome = *it;
        }
        if(it != all_palindromes.begin()){ // Check palindrome less than/equal to k
            it--;
            if(k - *it <= best){
                best = k - *it;
                best_palindrome = *it;
            }
        }
        cout << best_palindrome << '\n';
    }
}

File: Week 5/solutions/D/D_password.py
import bisect

INF = 10**9 + 5


# There are a lot of approaches to this problem; the constraints are small enough
# that you can iterate over all the palindromes to find the smallest. But it becomes
# simpler to write if we just precompute them, and use a list to find the smallest

# We can store all the palindromes in a sorted list
all_palindromes = []

for i in range(100, 1000): # No leading zeroes; start at 100
    s = str(i)
    t = s[::-1]
    s += t
    all_palindromes.append(int(s))

qr = int(input())
for _ in range(qr):
    k = int(input())
    # Find the first palindrome greater than k
    # bisect.bisect_left returns the index of the first element that is greater than or equal to x
    # bisect.bisect_right returns the index of the first element that is greater than x
    right = bisect.bisect_right(all_palindromes, k)
    best = INF
    best_palindrome = -1
    if right < len(all_palindromes): # Check palindrome greater than k
        best = all_palindromes[right] - k
        best_palindrome = all_palindromes[right]
    if right > 0: # Check palindrome less than/equal to k
        left = right - 1
        if k - all_palindromes[left] <= best:
            best = k - all_palindromes[left]
            best_palindrome = all_palindromes[left]
    print(best_palindrome)
